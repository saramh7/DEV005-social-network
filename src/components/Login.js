/**
 * Vista 1. Inicia sesión
 */
import {
  googleLogin, validateSession, loginWithUserEmail,
} from '../lib/auth';
import { loginTemplate } from '../templates/login.js';

function Login(navigateTo) {
  const loginContainer = document.createElement('div');
  loginContainer.classList.add('loginContainer');
  loginContainer.innerHTML = loginTemplate; // Inserta el template de templates/login.js

  /**
  * Por revisar, no está funcionando correctamente.
  * Se supone que debe validar si el usuario tiene una sesión activa para redirigirlo
  * directamente al Home
  */
  // const validate = validateSession();
  // if (validate) {
  //   console.log('va');
  //   navigateTo('/home');
  //   return '<div></div>';
  // }

  /**
  * Agrega evento click al botón de ingresar con google
  */
  const loginGoogleBtn = loginContainer.querySelector('#loginGoogleBtn');
  loginGoogleBtn.addEventListener('click', async () => {
    try {
      // Invoca función googleLogin del archivo lib/auth.js y recibe datos retornados
      const accessData = await googleLogin();
      if (accessData.loginOk) {
        navigateTo('/home'); // Si el ingreso del usuario es correcto lo redirige al Home.
      } else {
        const statusLogin = loginContainer.querySelector('#statusLogin');
        switch (accessData.data.code) {
          case 'auth/popup-closed-by-user':
            statusLogin.innerText = 'Lo sentimos, la ventana de ingreso fue cerrada.';
            break;
          default:
            statusLogin.innerText = 'Hubo un problema con tu inicio de sesión. Contacta al administrador de la App';
        }
      }
    } catch (error) {
      // Si algo falla en el proceso de login con Google muestra este log por consola y muestra
      // un error en el label id="statusLogin"
      console.warn('Login.js - loginGoogleBtn - click - Error: ', error);
      const statusLogin = loginContainer.querySelector('#statusLogin');
      statusLogin.innerText = 'Tuvimos un problema. Intente más tarde';
    }
  });

  /**
  * Agrega evento click al botón de ingresar con email y password
  */
  const loginBtn = loginContainer.querySelector('#loginBtn');
  loginBtn.addEventListener('click', async () => {
    // Lee los datos ingresados por el usuario en el campo email y password
    const email = loginContainer.querySelector('#email').value;
    const password = loginContainer.querySelector('#password').value;

    // Instancia el label de error del login (donde se mostrarán los errores)
    const statusLogin = loginContainer.querySelector('#statusLogin');

    // Valida que el usuario haya ingresado ambos datos
    if (email === '' || password === '') {
      statusLogin.innerText = 'Debe ingresar su email y password antes de continuar.';
      return null;
    }

    try {
      // Invoca función loginWithUserEmail del archivo lib/auth.js y recibe datos retornados
      const accessData = await loginWithUserEmail(email, password);
      if (accessData.loginOk) {
        navigateTo('/home'); // Si el ingreso del usuario es correcto lo redirige al Home.
      } else {
        // Si el ingreso del usuario es incorrecto muestra un mensaje de error
        // en el label id="statusLogin" según el error que se recibe desde Firebase
        // hay muchas posibilidades de error. Sólo estamos mostrando mensajes personalizados
        // para 2 posibles errores: user-not-found y wrong-password
        switch (accessData.data.code) {
          case 'auth/user-not-found':
            // link para documentación: https://firebase.google.com/docs/auth/admin/errors?hl=es-419
            statusLogin.innerText = 'Lo sentimos, no pudimos encontrar tu cuenta';
            break;
          case 'auth/wrong-password':
            statusLogin.innerText = 'La contraseña es inválida';
            break;
          default:
            statusLogin.innerText = 'Hubo un problema con tu inicio de sesión. Contacta al administrador de la App';
        }
      }
    } catch (error) {
      // Si algo falla en el proceso de login con Google muestra este log por consola y muestra
      // un error en el label id="statusLogin"
      console.warn('Login.js - loginBtn - click - Error: ', error);
      statusLogin.innerText = 'Tuvimos un problema. Intente más tarde';
    }

    return null;
  });
  return loginContainer;
}

export default Login;
