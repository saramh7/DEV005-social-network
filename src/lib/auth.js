import {
  GoogleAuthProvider, signInWithPopup, signOut,
} from 'firebase/auth';
import { auth } from './firebase.js';

const provider = new GoogleAuthProvider();
auth.languageCode = 'es';

export const googleLogin = async () => {
  await signInWithPopup(auth, provider)
    .then((result) => {
      // This gives you a Google Access Token. You can use it to access the Google API.
      const credential = GoogleAuthProvider.credentialFromResult(result);
      const token = credential.accessToken;
      // The signed-in user info.
      const user = result.user;
      // IdP data available using getAdditionalUserInfo(result)

      console.log('Usuario Loggeado exitosamente', result);
    }).catch((error) => {
      // Handle Errors here.
      const errorCode = error.code;
      const errorMessage = error.message;
      // The email of the user's account used.
      const email = error.customData.email;
      // The AuthCredential type that was used.
      const credential = GoogleAuthProvider.credentialFromError(error);
      // ...

      console.log('Error', error);
    });
};

export const googleLogout = async () => {
  await signOut(auth).then(() => {
    console.log('Sign-out successful');
    // Sign-out successful.
  }).catch((error) => {
    console.log('Error', error);
    // An error happened.
  });
};
