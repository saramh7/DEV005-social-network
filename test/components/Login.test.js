/**
 * @jest-environment jsdom
 */

import Login from '../../src/components/Login.js';
import * as auth from '../../src/lib/auth.js';

const navigateTo = jest.fn();

describe('googleLogin', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('Testing DOM Elements', () => {
    test('it should be a function', () => {
      expect(typeof Login).toBe('function');
    });

    test('it should have a login button', () => {
      const DOM = document.createElement('div');
      DOM.append(Login(navigateTo));
      const loginBtn = DOM.querySelector('#loginBtn');
      expect(loginBtn).toBeTruthy();
    });

    test('it should have a login with google button', () => {
      const DOM = document.createElement('div');
      DOM.append(Login(navigateTo));
      const loginGoogleBtn = DOM.querySelector('#loginGoogleBtn');
      expect(loginGoogleBtn).toBeTruthy();
    });

    test('it should have an email and password inputs', () => {
      const DOM = document.createElement('div');
      DOM.append(Login(navigateTo));
      const email = DOM.querySelector('#email');
      const password = DOM.querySelector('#password');
      expect(email).toBeTruthy();
      expect(password).toBeTruthy();
    });

    test('it should have an Label for error description', () => {
      const DOM = document.createElement('div');
      DOM.append(Login(navigateTo));
      const label = DOM.querySelector('#statusLogin');
      expect(label).toBeTruthy();
    });
  });

  describe('Testing Login with Google', () => {
    test('it should call function googleLogin and redirect to home', () => {
      jest.spyOn(auth, 'googleLogin').mockImplementation(() => Promise.resolve({ loginOk: true, data: { code: 'mockData' } }));
      const DOM = document.createElement('div');
      DOM.append(Login(navigateTo));
      const loginGoogleBtn = DOM.querySelector('#loginGoogleBtn');
      loginGoogleBtn.click();
      expect(auth.googleLogin).toHaveBeenCalledTimes(1);
      setTimeout(() => {
        expect(navigateTo).toHaveBeenCalledTimes(1);
        expect(navigateTo).toHaveBeenCalledWith('/home');
      }, 0);
    });

    test('it should show an error when googleLogin return loginOk=false', () => {
      jest.spyOn(auth, 'googleLogin').mockImplementation(() => Promise.resolve({ loginOk: false, data: { code: 'auth/popup-closed-by-user' } }));
      const DOM = document.createElement('div');
      DOM.append(Login(navigateTo));
      const loginGoogleBtn = DOM.querySelector('#loginGoogleBtn');
      const statusLogin = DOM.querySelector('#statusLogin');
      loginGoogleBtn.click();
      setTimeout(() => {
        expect(statusLogin.innerText).toBe('Lo sentimos, la ventana de ingreso fue cerrada.');
      }, 0);
    });

    test('it should show an error when googleLogin return another kind of error', () => {
      jest.spyOn(auth, 'googleLogin').mockImplementation(() => Promise.resolve({ loginOk: false, data: { code: 'auth/another-error' } }));
      const DOM = document.createElement('div');
      DOM.append(Login(navigateTo));
      const loginGoogleBtn = DOM.querySelector('#loginGoogleBtn');
      const statusLogin = DOM.querySelector('#statusLogin');
      loginGoogleBtn.click();
      setTimeout(() => {
        expect(statusLogin.innerText).toBe('Lo sentimos, la ventana de ingreso fue cerrada.');
      }, 0);
    });

    test('it should show an error when googleLogin return a promise rejected', async () => {
      jest.spyOn(auth, 'googleLogin').mockImplementation(() => Promise.reject());
      const DOM = document.createElement('div');
      DOM.append(Login(navigateTo));
      const loginGoogleBtn = DOM.querySelector('#loginGoogleBtn');
      const statusLogin = DOM.querySelector('#statusLogin');
      loginGoogleBtn.click();
      setTimeout(() => {
        expect(statusLogin.innerText).toBe('Tuvimos un problema. Intente más tarde');
      }, 0);
    });
  });

  describe('Testing Login with Email and Password', () => {
    test('it should show error message when try login without entering email or password', () => {
      const DOM = document.createElement('div');
      DOM.append(Login(navigateTo));
      const loginBtn = DOM.querySelector('#loginBtn');
      loginBtn.click();
      const statusLogin = DOM.querySelector('#statusLogin');
      expect(statusLogin.innerText).toBe('Debe ingresar su email y password antes de continuar.');
    });

    test('it should show call loginWithUserEmail and redirect to home when loginWithUserEmail return loginOk=true', () => {
      jest.spyOn(auth, 'loginWithUserEmail').mockImplementation(() => Promise.resolve({ loginOk: true, data: { code: 'mockData' } }));
      const DOM = document.createElement('div');
      DOM.append(Login(navigateTo));
      const loginBtn = DOM.querySelector('#loginBtn');
      DOM.querySelector('#email').value = 'test@test.com';
      DOM.querySelector('#password').value = '123456';
      loginBtn.click();
      expect(auth.loginWithUserEmail).toHaveBeenCalledTimes(1);
      expect(auth.loginWithUserEmail).toHaveBeenCalledWith('test@test.com', '123456');
      setTimeout(() => {
        expect(navigateTo).toHaveBeenCalledTimes(1);
        expect(navigateTo).toHaveBeenCalledWith('/home');
      }, 0);
    });

    test('it should show an error when loginWithUserEmail return loginOk=false with code auth/user-not-found', async () => {
      jest.spyOn(auth, 'loginWithUserEmail').mockImplementation(() => Promise.resolve({ loginOk: false, data: { code: 'auth/user-not-found' } }));
      const DOM = document.createElement('div');
      DOM.append(Login(navigateTo));
      const loginBtn = DOM.querySelector('#loginBtn');
      DOM.querySelector('#email').value = 'test@test.com';
      DOM.querySelector('#password').value = '123456';
      const statusLogin = DOM.querySelector('#statusLogin');
      loginBtn.click();
      expect(auth.loginWithUserEmail).toHaveBeenCalledTimes(1);
      setTimeout(() => {
        expect(navigateTo).toHaveBeenCalledTimes(0);
        expect(statusLogin.innerText).toBe('Lo sentimos, no pudimos encontrar tu cuenta.');
      }, 0);
    });

    test('it should show an error when loginWithUserEmail return loginOk=false with another kind of error', async () => {
      jest.spyOn(auth, 'loginWithUserEmail').mockImplementation(() => Promise.resolve({ loginOk: false, data: { code: 'auth/another-error' } }));
      const DOM = document.createElement('div');
      DOM.append(Login(navigateTo));
      const loginBtn = DOM.querySelector('#loginBtn');
      DOM.querySelector('#email').value = 'test@test.com';
      DOM.querySelector('#password').value = '123456';
      const statusLogin = DOM.querySelector('#statusLogin');
      loginBtn.click();
      expect(auth.loginWithUserEmail).toHaveBeenCalledTimes(1);
      setTimeout(() => {
        expect(navigateTo).toHaveBeenCalledTimes(0);
        expect(statusLogin.innerText).toBe('Lo sentimos, no pudimos encontrar tu cuenta.');
      }, 0);
    });

    test('it should show an error when loginWithUserEmail return loginOk=false with code auth/wrong-password', async () => {
      jest.spyOn(auth, 'loginWithUserEmail').mockImplementation(() => Promise.resolve({
        loginOk: false,
        data: { code: 'auth/wrong-password' },
      }));
      const DOM = document.createElement('div');
      DOM.append(Login(navigateTo));
      const loginBtn = DOM.querySelector('#loginBtn');
      DOM.querySelector('#email').value = 'test@test.com';
      DOM.querySelector('#password').value = '123456';
      const statusLogin = DOM.querySelector('#statusLogin');
      loginBtn.click();
      expect(auth.loginWithUserEmail).toHaveBeenCalledTimes(1);
      setTimeout(() => {
        expect(navigateTo).toHaveBeenCalledTimes(0);
        expect(statusLogin.innerText).toBe('Lo sentimos, no pudimos encontrar tu cuenta.');
      }, 0);
    });

    test('it should show an error when loginWithUserEmail return a promise rejected', async () => {
      jest.spyOn(auth, 'loginWithUserEmail').mockImplementation(() => Promise.reject());
      const DOM = document.createElement('div');
      DOM.append(Login(navigateTo));
      const loginBtn = DOM.querySelector('#loginBtn');
      DOM.querySelector('#email').value = 'test@test.com';
      DOM.querySelector('#password').value = '123456';
      const statusLogin = DOM.querySelector('#statusLogin');
      loginBtn.click();
      expect(auth.loginWithUserEmail).toHaveBeenCalledTimes(1);
      setTimeout(() => {
        expect(navigateTo).toHaveBeenCalledTimes(0);
        expect(statusLogin.innerText).toBe('Tuvimos un problema. Intente más tarde');
      }, 0);
    });
  });
});
