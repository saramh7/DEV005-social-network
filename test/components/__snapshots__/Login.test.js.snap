// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loginWithUserEmail snapshot of googleLogin 1`] = `
<div>
  [object Promise]
</div>
`;
